Fix some Python 3 issues flagged by the -3 flag and Pylint

rowillia opened the pull request:
I've been working through looking at adopting Python 3, and I've been utilizing the `-3` flag to help idenitfy issues with Python 3.

Here's a quick write up I did last week:

https://gist.github.com/rowillia/c0feed97c1863b2d8e5a3ed73712df65

The `-3` flagged a few issues in `werkzeug` worth fixing, namely around `__eq__` shadowing `__hash__` in PY3.

Fixing these issues will help make the `-3` flag less noisey for folks running Werkzeug in production.

untitaker commented on werkzeug/contrib/atom.py:
```diff
@@ -159,7 +159,7 @@ def generate(self):
         # atom demands either an author element in every entry or a global one
         if not self.author:
-            if False in map(lambda e: bool(e.author), self.entries):
+            if any(not bool(e.author) for e in self.entries):
```
In that case `bool(..)` is not necessary.

rowillia replied:
Done

untitaker commented on werkzeug/datastructures.py:
```diff
@@ -765,6 +759,8 @@ def __eq__(self, other):
                 return False
         return True
 
+    __hash__ = None
```
I don't understand why this is necessary.

RonnyPfannschmidt replied:
i recall its a misfeature with eq vs hash - as far as i understood if you specify eq, hash wont be inherited from object
and to properly behave one needs to set it up

RonnyPfannschmidt replied:
https://docs.python.org/2/reference/datamodel.html#object.__hash__

rowillia commented:
@untitaker Thanks for taking a look!

I added some tests.  I added a pylint checker for this particular error...when they ship their next version it might be good to use their Python 3 compat checkers - https://github.com/PyCQA/pylint/pull/1115

rowillia commented:
@untitaker ping!  Anything else I can fix up here?

untitaker commented on tests/test_datastructures.py:
```diff
@@ -526,6 +526,12 @@ def test_ordered_interface(self):
+        d = self.storage_class()
+        d.add('foo', 23)
+        with pytest.raises(TypeError):
+            {d: 42}
```
`hash(d)` seems more readable to me

rowillia replied:
Done

RonnyPfannschmidt commented on tests/test_routing.py:
```diff
@@ -358,6 +358,13 @@ def test_rule_emptying():
+    rule = r.Rule('/foo', {'meh': 'muh'}, 'x', ['POST'],
+                  False, 'x', True, None)
+    with pytest.raises(TypeError):
+        {rule: 42}
```
dito

rowillia replied:
Done

rowillia commented:
@untitaker @RonnyPfannschmidt thanks for the feedback.  Fixed.

untitaker commented:
yeah lgtm iff ronny says it's conceptually good

On Fri, Oct 14, 2016 at 10:10:57AM -0700, Roy Williams wrote:
> @untitaker @RonnyPfannschmidt thanks for the feedback.  Fixed.
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/pallets/werkzeug/pull/1017#issuecomment-253862970

rowillia commented:
@RonnyPfannschmidt Anything else I can fix here?  CI failure seems unrelated.

RonnyPfannschmidt commented:
just missed the last comment, LGTM

untitaker commented:
Thanks!

rowillia commented:
Thanks @RonnyPfannschmidt and @untitaker !
